### PR TITLE
Implement Python API for launching asset backfills

### PR DESCRIPTION
**Pull request by deepyaman**
### Background

@smackesey has interacted with several users on the following scenario:

- There is some computation (run or backfill) X that should only be launched under certain conditions
- The conditions under which X should be launched are non-trivial to compute
- It is desired to manually trigger conditional launch of X-- that is, on clicking a button, the conditions under which X should be launched are computed and X is launched if those conditions are satisfied.

### Problem Statement

We do not have a good solution to the above scenario.

- Launch job Y that conditionally launches X
    - If X is a job, execute_job will work.
    - If X is a backfill, then there is no good public API to launch it. User must use the private GQL API.
- Manually trigger sensor that conditionally launches X
    - The way you manually trigger a sensor is by pressing the "Test Sensor" button, which is non-ideal and kinda buried.
    - You then have to pass each generated RunRequest through to the launchpad and manually launch.
        - This works OK if X is a job.
        - This is unworkable if X is a backfill, because you have to manually launch each RunRequest

So if X is a backfill, basically your only option is to use a job that then uses the private GQL API.

### Objective

The purpose of this enhancement is to improve user experience. This will allow users to not have to resort to workarounds to perform a manual trigger a conditional launch of X.

The manifestation of this is to expose a public Python API for creating and launching asset backfills.

## How I Tested These Changes

Added unit tests mirroring a couple of existing backfill tests, but using the newly-added Python API.

## TODO

I wanted to get feedback/validation that I'm on the right track before proceeding much further, but I understand that the following things probably still need doing:

- [ ] Test validation of arguments/argument combinations
- [ ] Add additional execution tests?
- [ ] Improve docstring (with examples, etc.)
- [ ] Add public-facing docs?

## Changelog

Exposed a Python API for creating and launching asset backfills.

### CODE REVIEW

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py [hunk 1]
```diff
@@ -114,11 +114,14 @@ def create_and_launch_partition_backfill(
...
-        asset_selection or backfill_params.get("selector") or backfill_params.get("partitionNames")
+        asset_selection
+        or backfill_params.get("selector")
+        or backfill_params.get("partitionNames")
+        or backfill_params.get("allPartitions")
     ) and partitions_by_assets:
         raise DagsterInvariantViolationError(
-            "partitions_by_assets cannot be used together with asset_selection, selector, or"
-            " partitionNames"
+            "partitionsByAssets cannot be used together with assetSelection, selector, or"
```
Comment (deepyaman) [inline python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:123]: At the very least, I feel like the capitalization should be consistent; the camel-casing is aligned with the GraphQL API names.

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py [hunk 2]
```diff
@@ -114,11 +114,14 @@ def create_and_launch_partition_backfill(
...
+            " partitionNames or if allPartitions is True"
```
Comment (deepyaman) [inline python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:124]: I think it makes sense to validate `allPartitions`, too, if checking `partitionNames`; neither can be used with `partitionsByAssets`.

I would be happy to split this and the above change into a separate (stacked?) PR, since they're not directly related to creating the Python API; just thought I'd do some housekeeping as I was learning about it for the Python API implementation.

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py [hunk 3]
```diff
@@ -231,10 +234,10 @@ def create_and_launch_partition_backfill(
         )
 
         backfill = PartitionBackfill.from_asset_partitions(
-            asset_graph=asset_graph,
             backfill_id=backfill_id,
-            tags=tags,
+            asset_graph=asset_graph,
             backfill_timestamp=backfill_timestamp,
+            tags=tags,
```
Comment (deepyaman) [inline python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:240]: Nit: Minimal argument order alignment across calls.

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 4]
```diff
@@ -2260,7 +2262,7 @@ def get_event_tags_for_asset(
     @public
     @traced
     def wipe_assets(self, asset_keys: Sequence[AssetKey]) -> None:
-        """Wipes asset event history from the event log for the given asset keys.
+        """Wipe asset event history from the event log for the given asset keys.
```
Comment (deepyaman) [inline python_modules/dagster/dagster/_core/instance/__init__.py:2265]: Nit: All of the other public methods in this file use imperative mood.

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 5]
```diff
@@ -3161,6 +3163,103 @@ def add_backfill(self, partition_backfill: "PartitionBackfill") -> None:
     def update_backfill(self, partition_backfill: "PartitionBackfill") -> None:
         self._run_storage.update_backfill(partition_backfill)
 
+    @public
+    def launch_backfill(
+        self,
+        asset_graph: "BaseAssetGraph",
```
Comment (deepyaman) [inline python_modules/dagster/dagster/_core/instance/__init__.py:3169]: I decided to directly ask for the asset graph as an argument, since there are already tutorials that ask the user to pass asset graph. I didn't really see any reason to ask for a broader `Definitions` object.

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 6]
```diff
@@ -3161,6 +3163,103 @@ def add_backfill(self, partition_backfill: "PartitionBackfill") -> None:
...
+            asset_selection (Optional[Sequence[AssetKey]]): List of asset keys to backfill.
+            partitions_by_assets (Optional[Sequence[Tuple[AssetKey, Optional[Sequence[Tuple[str, str]]]]]]):
+                List of partitions for each asset key to backfill.
+            partition_names (Optional[Sequence[str]]): List of partition names to backfill. Only
+                valid if `asset_selection` is provided.
+
+        Returns:
+            str: The ID of the backfill.
+        """
+        # TODO(deepyaman): Abstract logic shared with `dagster-graphql`.
```
Comment (deepyaman) [inline python_modules/dagster/dagster/_core/instance/__init__.py:3195]:
```suggestion
```

Actually, as I got to implementing more of this, maybe they're sufficiently different that this isn't worth it.

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 7]
```diff
@@ -3161,6 +3163,103 @@ def add_backfill(self, partition_backfill: "PartitionBackfill") -> None:
...
         self._run_storage.update_backfill(partition_backfill)
 
+    @public
+    def launch_backfill(
+        self,
+        asset_graph: "BaseAssetGraph",
+        asset_selection: Optional[Sequence[AssetKey]] = None,
+        partitions_by_assets: Optional[
+            Sequence[Tuple[AssetKey, Optional[Sequence[Tuple[str, str]]]]]
+        ] = None,
```
Comment (deepyaman) [inline python_modules/dagster/dagster/_core/instance/__init__.py:3173]: This avoids exposing any additional objects, but I suppose an explanation/examples will be necessary for anybody unfamiliar with Dagster internals already.

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py [hunk 8]
```diff
@@ -114,17 +114,22 @@ def create_and_launch_partition_backfill(
...
-            " partitionNames"
+            "partitionsByAssets cannot be used together with assetSelection, selector, or"
+            " partitionNames or if allPartitions is True"
         )
 
     tags = {t["key"]: t["value"] for t in backfill_params.get("tags", [])}
 
     tags = {**tags, **graphene_info.context.get_viewer_tags()}
 
+    title = check_valid_title(backfill_params.get("title"))
```
Comment (deepyaman) [inline python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:131]: It seems reasonable to use the existing utility. Also, I think there's no reason it should only be happening for job backfills.

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 9]
```diff
@@ -3161,6 +3164,122 @@ def add_backfill(self, partition_backfill: "PartitionBackfill") -> None:
...
+                title=title,
+                description=description,
+            )
+        else:
+            raise DagsterError(
+                "Backfill requested without specifying partition set selector or asset selection"
+            )
+
+        # Check that the partition keys can be found in the asset graph.
+        # TODO(deepyaman): Abstract logic shared with `dagster-graphql`.
```
Comment (deepyaman) [inline python_modules/dagster/dagster/_core/instance/__init__.py:3261]: I think this one can definitely be deduplicated; probably can be done in a subsequent PR?

**Review by OwenKephart — CHANGES_REQUESTED** [review]
After sleeping on this, I feel like putting this logic on the DagsterInstance itself isn't quite right.

At the core, there are two things we want to be able to do here:

1. Express some intent to backfill some arbitrary set of partitions
2. Send a message to the instance to act on that intent

We already have an existing method for (2) (instance.add_backfill), and so the question is what is the best way of exposing (1) to the user.

We already have the `PartitionBackfill` object, and so putting this layer of indirection for constructing the partition backfill feels off, especially because none of the code in the added method really interacts with the core instance database outside of a single call to self.add_backfill().

Right now, this single method supports a wide array of different parameter combinations, with only some subset of them being valid, which is pretty confusing for a user, even if we add a ton of examples of valid behavior.

I think the better way of factoring this behavior would be to just expose PartitionBackfill more directly, and make the existing static constructors public (with a potentially a bit of light refactoring of the input types, e.g. backfill_id could be an optional parameter -- if not provided, we'd create one for the user).

To start, we could just expose `.from_asset_partitions()` as public, as I think that's the most common case, and its existing implementation is less subject to change than the from_partitions_by_assets (which has extra custom types)

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py (hunk 8 above)
Comment (jamiedemaria) [reply python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:131]: doing it for only job backfills is 100% just a goof on my part. we should be doing it for all types of backfills - thanks for catching! 

switching to the check version of the fn is also fine w me. I think originally we were considering re-raising as a GrapheneError so that it had a more immediate failure for the user. But we've kind of abandoned adding titles/descriptions to backfills for now anyway, so keeping it as the generic python error is fine

**Comment by jamiedemaria** [comment]
++1 I think exposing the constructors on `PartitionBackfill` is a lot cleaner and gives us more flexibility in the future

**Review by jamiedemaria — CHANGES_REQUESTED** [review]
Have there been discussions in the build team about the implications of allowing users to launch backfills (and other jobs) from runs? I know in the past we've shied away from allowing/promoting this behavior and instead pointed users to sensors. Would love to get up to date on what you are all thinking in terms of best practices here

File: python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py (hunk 8 above)
Comment (deepyaman) [reply python_modules/dagster-graphql/dagster_graphql/implementation/execution/backfill.py:131]: @jamiedemaria Great! I raised a new PR just addressing this (tagged you, if you can take a quick look), since it's agreed upon, while I go about changing this broader PR.

**Comment by gibsondan** [comment]
We should make sure to consider how this will work in cloud (where the instance api needs to communicate over graphql) and what the exact api boundaries will be that we need to preserve backwards compatibility for

**Comment by deepyaman** [comment]
> After sleeping on this, I feel like putting this logic on the DagsterInstance itself isn't quite right.
> 
> At the core, there are two things we want to be able to do here:
> 
> 1. Express some intent to backfill some arbitrary set of partitions
> 2. Send a message to the instance to act on that intent
> 
> We already have an existing method for (2) (instance.add_backfill), and so the question is what is the best way of exposing (1) to the user.
> 
> We already have the `PartitionBackfill` object, and so putting this layer of indirection for constructing the partition backfill feels off, especially because none of the code in the added method really interacts with the core instance database outside of a single call to self.add_backfill().
> 
> Right now, this single method supports a wide array of different parameter combinations, with only some subset of them being valid, which is pretty confusing for a user, even if we add a ton of examples of valid behavior.
> 
> I think the better way of factoring this behavior would be to just expose PartitionBackfill more directly, and make the existing static constructors public (with a potentially a bit of light refactoring of the input types, e.g. backfill_id could be an optional parameter -- if not provided, we'd create one for the user).
> 
> To start, we could just expose `.from_asset_partitions()` as public, as I think that's the most common case, and its existing implementation is less subject to change than the from_partitions_by_assets (which has extra custom types)

I should have a separate PR up later today just exposing `PartitionBackfill.from_asset_partitions()`! 🤞

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 10]
```diff
@@ -3261,8 +3345,8 @@ def get_latest_materialization_code_versions(
 
         return result
 
-    @experimental
     @public
+    @experimental
```
Comment (deepyaman) [inline python_modules/dagster/dagster/_core/instance/__init__.py:3349]: Minor consistency thing; feels like everywhere else in the codebase `@public` is above `@experimental`.

**Comment by deepyaman** [comment]
And we're back (from #26786)! After discussion with @OwenKephart, this updated PR takes a middle ground, only supporting backfills from asset partitions (no partitions by assets), but exposing a limited API rather than any objects used internally.

In the future, it may still make sense to expose something like `EntitySubset` (under a name TBD), but this PR is implemented such that that could be added as an argument to `launch_backfill` and the other options could potentially be deprecated over time (we expect the exposed options to be, at their core, still necessary to construct a backfill in the future; however, the specific API could change). To that end, the API is also marked experimental, and all parameters are keyword-only to avoid potential backward compatibility risks to the extent possible, if and when other arguments should be introduced.

**Comment by gibsondan** [comment]
> > We should make sure to consider how this will work in cloud (where the instance api needs to communicate over graphql) and what the exact api boundaries will be that we need to preserve backwards compatibility for
> 
> Is this still a significant concern with the intentionally-narrowed API?

not neccesarily, a blocking concern, but definitely something to consider the implications of. It looks like we technically already have a graphql mutation for add_backfill: https://github.com/dagster-io/internal/blob/master/dagster-cloud/python_modules/dagster-cloud/dagster_cloud/storage/runs/storage.py#L544-L549 but I doubt it was really getting much if any use.

Once this becomes a fully blessed/public thing,wWhat that means is that we need to make sure that any code that reads these backfill objects (e.g. the backfill daemon) can also handle old serialized objects as well that might have originated from earlier versions.

File: python_modules/dagster/dagster/_core/instance/__init__.py [hunk 11]
```diff
@@ -3161,6 +3164,87 @@ def add_backfill(self, partition_backfill: "PartitionBackfill") -> None:
...
+                asset_graph=asset_graph,
+                asset_selection=asset_selection,
+                partition_names=partition_names,
+                backfill_timestamp=backfill_timestamp,
+                tags=tags,
+                dynamic_partitions_store=dynamic_partitions_store,
+                all_partitions=all_partitions,
+                title=title,
+                description=description,
+            )
```
Comment (gibsondan) [inline python_modules/dagster/dagster/_core/instance/__init__.py:3219]: from my perspective it would be really nice if as much of this as possible can be done on the 'server' (which isn't really a thing in our instance API unfortunately...). Right now this is a lot of asset graph calculation happening on the 'client', at least in cloud

It's really too bad that the asset graph is not serialized in the database in OSS currently / needs to be passed in by the caller. That is not the case in cloud and simplifies the API in situations like this quite a bit i think

**Review by gibsondan — COMMENTED** [review]
Thinking about this more, it doesn't seem to me like you should have to pass in the asset graph here in this new API... the daemon that is going to read in this object and try to actually launch the backfill already has access to that asset graph and is capable of doing the same calculation that is happening here in the launch_backfill method.

Is there a world where we can defer more of that computation to the daemon instead? I appreciate that that makes things more complicated (would require bigger changes at both the write layer and the read layer to defer that computation) but I think it makes the user facing API simpler and becomes much less gnarly to handle back-compat for this in cloud

File: python_modules/dagster/dagster/_core/instance/__init__.py (hunk 11 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/instance/__init__.py:3219]: Is there a particular goal in pushing this work onto the "server" side in cloud? My main concern here is that if we try to pass in some sort of "unresolved" asset graph subset selection (i.e. some combination of parameters such as asset_selection, partition_names, all_partitions), then this is actually a lot less flexible as time goes on. We'd need to commit to those parameters being the "official" way of determining an asset graph subset, when in reality I think we'll want to evolve those python APIs over time.

Concretely, I'm imagining in the mid-term future we'll want to start exposing `EntitySubset`-type APIs more publicly, which would mean that we'd want to add a parameter to this method that accepts a list of EntitySubsets determining what to backfill.

In that future, it's much easier to rely on PartitionBackfill as the serialized object that we produce here, as that is maximally flexible -- we can construct that object in whatever way we want on the client side, and be confident that the daemon can read it.

**Comment by OwenKephart** [comment]
@gibsondan 

re: "it doesn't seem to me like you should have to pass in the asset graph here in this new API."

To summarize my comment above, my main feeling here is that it's simpler in the GQL layer if we rely on the PartitionBackfill object as the serialized interface between the "client" and "server" in cloud, as this means we don't have to update any GQL code as we eventually evolve and improve our APIs for specifying AssetGraphSubsets in user code

**Comment by gibsondan** [comment]
Will you be able to launch a backfill that involves assets from multiple code locations though? Seems like today the answer is no, since you will only have access to the subset of the asset graph that is from the calling code location? Or is that what you are getting it re: "evolving and improving our APIs for specifying AssetGraphSubsets"?

File: python_modules/dagster/dagster/_core/instance/__init__.py (hunk 5 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/instance/__init__.py:3172]: Can you point me to the tutorial where you see this? I'm having trouble finding one other than the airlift tutorial, which is a slightly different scenario since it is getting the asset graph from `defs` in a `__main__` not in the body of a job

**Comment by jamiedemaria** [comment]
One thing that would be useful for me to see is some examples of how/where we expect users to use this API. I'm a bit concerned that we are adding a third way to do a similar type of operation (launch some executions) - we already have the `execute_job` API which isn't called via the instance, and emitting `RunRequests` from sensors. Adding an instance method for launching a backfill is a bit of a departure from both of these apis. I think it could be strange to explain to the user "if you want to launch a run from a sensor, use RunRequest, but if you want to launch a backfill, use instance.launch_backfill", or "if you want to launch a run from another run, use `execute_job` but if you want to launch a backfill use `instance.launch_backfill`". 

Also potentially strange, but maybe fine, is that if you were to call `execute_job` from another job, the top level job waits for the called job to complete before continuing (at least in my testing, but maybe i'm missing a setting), but `instance.launch_backfill` would create the backfill, but not wait for it to execute before continuing. I'm not convinced that this is an issue, and i definitely don't think we should be waiting for the backfill to complete before continuing with the job, just noting that it is a behavior difference that we should at least be aware of for writing docs. 

My final note is that the current instance method doesn't have a path for launching a job-backfill. I think this is probably fine, since job backfills are kind of a strange code path anyway and you can get equivalent behavior by just launching a run per partition of the job in a loop. Maybe this is where `execute_job` actually running the full job and not just submitting it to the run coordinator becomes a bit more of a problem though 

Overall though, i do definitely prefer this simplified version of the method. It gives a lot less confusion on what the valid parameter combos are

**Comment by deepyaman** [comment]
> One thing that would be useful for me to see is some examples of how/where we expect users to use this API. I'm a bit concerned that we are adding a third way to do a similar type of operation (launch some executions) - we already have the `execute_job` API which isn't called via the instance, and emitting `RunRequests` from sensors. Adding an instance method for launching a backfill is a bit of a departure from both of these apis. I think it could be strange to explain to the user "if you want to launch a run from a sensor, use RunRequest, but if you want to launch a backfill, use instance.launch_backfill", or "if you want to launch a run from another run, use `execute_job` but if you want to launch a backfill use `instance.launch_backfill`".
> 
> Also potentially strange, but maybe fine, is that if you were to call `execute_job` from another job, the top level job waits for the called job to complete before continuing (at least in my testing, but maybe i'm missing a setting), but `instance.launch_backfill` would create the backfill, but not wait for it to execute before continuing. I'm not convinced that this is an issue, and i definitely don't think we should be waiting for the backfill to complete before continuing with the job, just noting that it is a behavior difference that we should at least be aware of for writing docs.
> 
> My final note is that the current instance method doesn't have a path for launching a job-backfill. I think this is probably fine, since job backfills are kind of a strange code path anyway and you can get equivalent behavior by just launching a run per partition of the job in a loop. Maybe this is where `execute_job` actually running the full job and not just submitting it to the run coordinator becomes a bit more of a problem though

I think one of the things that discussed (but maybe didn't properly document) when picking up this ticket was that it's not _quite_ like `execute_job`, in that it's adding it to get picked up by the daemon instead of actually triggering it (e.g. by doing something like `execute_backfill_iteration_loop`).

I think `DagsterInstance.launch_*` is at least somewhat consistent with other `DagsterInstance` methods FWIW.

**Comment by jamiedemaria** [comment]
> it's not quite like execute_job, in that it's adding it to get picked up by the daemon instead of actually triggering it (e.g. by doing something like execute_backfill_iteration_loop).

that gets to my point pretty well because at the surface, these two methods solve similar problems (i want to launch a run and i want to launch a backfill), but they use different APIs and have different side effects. 

> I think DagsterInstance.launch_* is at least somewhat consistent with other DagsterInstance methods FWIW.

Do users actually call `instance.launch_run` though? it's not a public, and even if you were to use it, you'd have to create the run first so that you can pass the id to the method, which is a pretty different user experience

I don't feel strongly that either of these are reasons to abandon this approach, just things to keep in mind and make sure we are comfortable adding to the api

File: python_modules/dagster/dagster/_core/instance/__init__.py (hunk 5 above)
Comment (deepyaman) [reply python_modules/dagster/dagster/_core/instance/__init__.py:3172]: It was an Airlift tutorial. 😅

**Comment by OwenKephart** [comment]
Just summarizing my take here to address @jamiedemaria and @gibsondan's comments:

1. I think we definitely need some programmatic way of sending compute over to the backfill daemon from python. This requires some set of parameters (tags, description, etc.), as well as some way of expressing "these asset partitions should be kicked off".
2. We'll need some sort of instance method for this purpose, and I think having this be "async" by default (i.e. it does not block on backfill completion) is the most natural -- generally the launching process will just want to express an intent to materialize stuff. You can always write a loop to poll for completion of the returned backfill id if you so choose. I agree that this is a bit different in feel from job-based APIs but this is probably ok, just because the operation you're doing is pretty different here
3. As time goes on, we'll definitely improve our APIs for expressing subsets of an asset graph -- those methods might end up not include passing in an AssetGraph object at all, depending on how we do this. We can adapt the parameters to this method to allow for those new methods to work seamlessly with this API (potentially going as far as to hide the parameters we're currently working with)

Basically I think this API is a reasonable starting point and doesn't inhibit our options much in the future

@jamiedemaria to your point about not being able to do job backfills from this method, maybe this should be called `launch_asset_backfill`?

**Comment by gibsondan** [comment]
@OwenKephart i'm still thrown by the fact that the asset graph that is being passed here is different from the one that the backfill daemon is actually going to use (in that it will be repository scoped). Is that going to limit the set of things that you can express with this API compared to what you could express over graphql?

**Comment by deepyaman** [comment]
* **#27060** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/27060?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#26731** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26731?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26731?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by deepyaman** [comment]
@schrockn Would be great to get your view, and maybe that can also help us move towards alignment. I think we've gone back and forth a bit on what a Python backfill API should look like, what objects should be exposed, and how it may evolve in the future, etc. (due to a number of valid questions/concerns raised).

**Review by schrockn — CHANGES_REQUESTED** [review]
1) The logic should absolutely not be on `DagsterInstance`. It is already a bloated [God class](https://github.com/dagster-io/internal/discussions/9791), and most of the "public APIs" on the class are accidental ones. Users should be able to launch runs and or Backfilsls without newing up a instance.

2) We should definitely consolidate whatever implementation is in the the graphQL server with this. If the implementations don't stack, that is itself a pretty serious. They should stack on one another. I'd like to understand why.

3) I do not believe we have ever used `BaseAssetGraph` or `AssetGraph` as an input to one our public APIs. However we certainly need to have an abstraction to represent an arbitrary sets of partitions that possibly spans multiple asset keys. I doubt AssetGraph is the right one.

4) When considering any API choice here (there's a lot of discussion here) we need to leave ourselves option value here for a world where we move more of the backfill daemon logic to user processes.

My recommendation for designing this API is a README-driven development process. Write the doc that we would point a user to if we were asked "How do I launch a backfill from Python".

**Comment by deepyaman** [comment]
> 1. The logic should absolutely not be on `DagsterInstance`. It is already a bloated [God class](https://github.com/dagster-io/internal/discussions/9791), and most of the "public APIs" on the class are accidental ones. Users should be able to launch runs and or Backfilsls without newing up a instance.

Makes sense. In the README recommended below, I'll make sure to keep this in mind.

> 2. We should definitely consolidate whatever implementation is in the the graphQL server with this. If the implementations don't stack, that is itself a pretty serious. They should stack on one another. I'd like to understand why.

I think the logic/coverage can be essentially identical; the very first commit in this PR (https://github.com/dagster-io/dagster/pull/26731/commits/aa12a4520b9377bdb18671552f5af38303be42c3) is as close to a direct port of the GraphQL logic (excluding job backfills) that I could get. Even now, the logic is aligned; it's only that the API (intentionally) only exposes the most common case of asset backfill from partitions.

> 3. I do not believe we have ever used `BaseAssetGraph` or `AssetGraph` as an input to one our public APIs. However we certainly need to have an abstraction to represent an arbitrary sets of partitions that possibly spans multiple asset keys. I doubt AssetGraph is the right one.

I think @OwenKephart mentioned needing a new concept like `EntitySubset`—please correct me if I'm misremembering—that doesn't exist yet/would need to be designed; I think this is quite aligned with what you're mentioning here.

> 4. When considering any API choice here (there's a lot of discussion here) we need to leave ourselves option value here for a world where we move more of the backfill daemon logic to user processes.

Will need to understand this more myself; don't think I have the context on what this future world might look like.

> My recommendation for designing this API is a README-driven development process. Write the doc that we would point a user to if we were asked "How do I launch a backfill from Python".

I'll try to tackle this first and have something by tomorrow or Monday.

**Comment by deepyaman** [comment]
> My recommendation for designing this API is a README-driven development process. Write the doc that we would point a user to if we were asked "How do I launch a backfill from Python".

I've drafted a README for launching backfills in #27260. The main differences are:

> The logic should absolutely not be on `DagsterInstance`.

`DagsterInstance` is just passed in, derived from `context`.

> I do not believe we have ever used `BaseAssetGraph` or `AssetGraph` as an input to one our public APIs.

I still kept an `asset_graph` kwarg, but the user is getting it from `context.repository_def.asset_graph`. If it's better/more familiar to just pass `repository_def`, that would also work.

Focusing on the motivating use cases (launching from a sensor or schedule) helps mitigate some of the new/contentious concepts, since creating a `RunRequest` for launching a backfill is already supported—what's missing is a public way to construct that backfill, and the two methods proposed in #27260 provide ways to do that.

**Comment by jamiedemaria** [comment]
i'm going to close this for review queue management
